src/util/rust_util/sos_util.rs

```diff
@@ -0,0 +1,274 @@
+//! This module is a tool for visiting slice of slices and group adjacent elements by keys.
+//! It is useful for coalescing `mmap` calls for requests from `FragmentedMapper`.
+
+/// A group of adjacent elements in a slice of slices with the same key.
+pub(crate) struct Group<'s, T, K>
+where
+    K: PartialEq + Copy,
+{
+    slice_of_slices: &'s [&'s [T]],
+    /// The key of the elements in this group
+    pub(crate) key: K,
+    start_outer: usize,
+    start_inner: usize,
+    /// The starting index of the group among all elements in the slice of slices.
+    pub(crate) start_total: usize,
+    end_outer: usize,
+    end_inner: usize,
+    /// The ending index (exclusive) of the group among all elements in the slice of slices.
+    pub(crate) end_total: usize,
+}
+
+impl<'s, T, K> Group<'s, T, K>
+where
+    K: PartialEq + Copy,
+{
+    pub(crate) fn iter(&'s self) -> GroupIterator<'s, T, K> {
+        GroupIterator {
+            group: self,
+            cur_outer: self.start_outer,
+            cur_inner: self.start_inner,
+        }
+    }
+}
+
+/// Iterate elements in a group.
+pub(crate) struct GroupIterator<'g, T, K>
+where
+    K: PartialEq + Copy,
+{
+    group: &'g Group<'g, T, K>,
+    cur_outer: usize,
+    cur_inner: usize,
+}
+
+impl<'g, T, K> Iterator for GroupIterator<'g, T, K>
+where
+    K: PartialEq + Copy,
+{
+    type Item = &'g T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.cur_outer == self.group.end_outer && self.cur_inner == self.group.end_inner {
+            return None;
+        }
+
+        let rv = &self.group.slice_of_slices[self.cur_outer][self.cur_inner];
+        self.cur_inner += 1;
+        if self.cur_inner == self.group.slice_of_slices[self.cur_outer].len() {
+            self.cur_inner = 0;
+            self.cur_outer += 1;
+        }
+
+        Some(rv)
+    }
+}
+
+/// An iterator that interates through groups of elements in a slice of slice.
+///
+/// Each group contains a contiguous range of elements that may span multiple inner slices.
+/// All elements in a group share the same key.
+pub(crate) struct SliceOfSlicesGrouper<'s, T, K, F>
```

Comment:
It took me a while to understand what this does. I think essentially it does [`flatten()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flatten) on the 'slice of slice', and [`group_by()`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.group_by) (from `itertools`) to group the resulting slice by keys. 

I would suggest try using those library methods and see if they work. In that case, we do not need to maintain this type. If they do not work for some reason, I suggest just adding comments for this type to make it clear that it does both `flatten` and `group_by`.

Reply:
Yes.  But one problem with the `group_by` provided by `Itertools` is that it is a one-way iterator.  But our mmapper needs to traverse each group twice to coalesce `mmap` calls:

1. First it finds all `MapState` elements with the same state, in order to compute the size of the memory range with the same state.  This size will be passed to `mmap`.
2. Then after calling `mmap`, it will visit every `MapState` in this group again to transition the state to `Quarantined`.

`Itertools::group_by` does not allow me to do this.

Fortunately, `flatten()` still does what we need.  I removed this `sos_util` module, and replaced with a simpler `RevisitableGroupBy`.  It is similar to `Itertools::group_by`, but it pre-iterate through the group to find its length before letting the user iterate through the group again.  It also play nicely with `.flatten()`, and this greatly simplified the implementation.